app/src/main/java/ch/epfl/sweng/SDP/home/LeaderboardActivity.java

```diff
@@ -0,0 +1,307 @@
+        exit.setOnTouchListener(new View.OnTouchListener() {
+            @Override
+            public boolean onTouch(View view, MotionEvent event) {
+                switch (event.getAction()) {
+                    case MotionEvent.ACTION_DOWN:
+                        HomeActivity.pressButton(exit, context);
+                        break;
+                    case MotionEvent.ACTION_UP:
+                        HomeActivity.bounceButton(view, HomeActivity.MAIN_AMPLITUDE,
+                                HomeActivity.MAIN_FREQUENCY, context);
+                        launchActivity(HomeActivity.class);
+                        break;
+                    default:
+                }
+                return true;
+            }
+        });
+    }
+
+    /**
+     * Gets called when user entered a new search query.
+     * @param query new string to search
+     */
+    private void updateLeaderboard(String query) {
+        Database.INSTANCE.getReference("users")
+                .orderByChild("username").startAt(query).endAt(query + "\uf8ff")
+                .addListenerForSingleValueEvent(new ValueEventListener() {
+                    @Override
+                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
+                        processResponse(dataSnapshot.getChildren());
+                    }
+
+                    @Override
+                    public void onCancelled(@NonNull DatabaseError databaseError) {
+                        Log.d(TAG, FIREBASE_ERROR);
+                    }
+                });
+    }
+
+    /**
+     * Gets called when all data has been fetched from Firebase.
+     * Adds snapshots to a list, sorts it by trophies and converts players to LinearLayouts
+     * @param snapshots all players from firebase fulfilling the search query
+     */
+    private void processResponse(Iterable<DataSnapshot> snapshots) {
+        leaderboard.removeAllViews();
+        LinkedList<Player> players = new LinkedList<>();
+        for (DataSnapshot s : snapshots) {
+            Player temp = new Player((String) s.child("userId").getValue(),
+                    (String) s.child("username").getValue(),
+                    (Long) s.child("trophies").getValue());
+
+            players.add(temp);
+        }
+        Collections.sort(players);
+        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
+                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT);
+        layoutParams.setMargins(0, 20, 0, 0);
+        for (int i = 0; i < Math.min(10, players.size()); ++i) {
+            leaderboard.addView(players.get(i).toLayout(getApplicationContext(), i), layoutParams);
+        }
+    }
+
+    /**
+     * Helper class to manage and display data from Firebase.
+     */
+    private class Player implements Comparable {
+
+        private String userId;
+        private String username;
+        private Long trophies;
```

Comment:
This field can all be final, and probably Long can be long, which is better for performance

Reply:
How can these fields be final if i pass the values in the constructor?
I used Long instead of long because this way I can directly call Long.compare(Long) in the compare method below. Otherwise I would have to write my own compare method. What do you think would be better?

Reply:
You can declare them as finals and assign them inside the constructor, it will compile. For Long you're right!

Reply:
You're right, it compiled. Thanks.